File: ndstrim.c
/*	This file is part of NDSTrim.
	
	NDSTrim is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.
	
	NDSTrim is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	
	You should have received a copy of the GNU General Public License
	along with NDSTrim.  If not, see <http://www.gnu.org/licenses/>
	
	
	ROM size is available in four bytes at 0x80-0x83.
	Wifi data is stored on 136 bytes after ROM data.
	Filesize is checked to be at least 0x200 bytes to make sure it contains a DS cartridge header.
	Filesize is then checked to be at least the rom size+wifi to avoid errors.
	
	Sources:
	http://nocash.emubase.de/gbatek.htm
	http://forums.ds-xtreme.com/showthread.php?t=1964
	http://gbatemp.net/index.php?showtopic=44022
*/

#include <stdio.h>
#include <stdlib.h>

//#define DEBUG
#define BUFFER 1000000 //1 MB buffer size

//The rom size is located in four bytes at 0x80-0x83
//Using a struct is the only way to manually specify the number of bits
struct uint32 {
	unsigned data:32; //4*8
};

int main(int argc, char *argv[]) {
	if (argc < 2) {
		fprintf(stderr,"%s: Too few arguments.\n",argv[0]);
		fprintf(stderr,"%s: Usage: %s <input> [output]\n",argv[0],argv[0]);
		exit(1);
	}
	
	//Debug?
	char debug=0;
	#ifdef DEBUG
	debug=1;
	#endif
	
	//Open input
	#ifdef DEBUG
	printf("%s: Opening input file '%s'.\n",argv[0],argv[1]);
	#endif
	FILE *input;
	if ((input=fopen(argv[1],"rb")) == NULL) {
		fprintf(stderr,"%s: fopen() failed in file %s, line %d.\n",argv[0],__FILE__,__LINE__);
		fprintf(stderr,"%s: This is most likely because the input file doesn't exist.\n",argv[0],__FILE__,__LINE__);
		exit(1);
	}
	
	//Get input filesize
	fseek(input,0,SEEK_END);
	unsigned int filesize=ftell(input);
	#ifdef DEBUG
	printf("%s: Filesize: %d bytes.\n",argv[0],filesize);
	#endif
	
	//Check if rom is big enough to contain a DS cartridge header
	if (filesize <= 0x200) {
		fprintf(stderr,"%s: Error: '%s' is too small to contain a NDS cartridge header (corrupt rom?).\n",argv[0],argv[1]);
		exit(1);
	}
	
	//Read rom size
	fseek(input,0x80,SEEK_SET);
	struct uint32 romsize;
	fread(&romsize,sizeof(romsize),1,input);
	unsigned int saferomsize=romsize.data;
	int wifi_block=0;
	
	//Check if romsize is 0 (some commercial roms use zero, unknown reason)
	//These roms usually have the whole rom filled with stuff (unknown if it's garbage or useful stuff)
	//Known roms: 0040, 0132, 0192, 0318, and 0357 have this (All these are Japanese roms)
	if (romsize.data == 0) {
		fprintf(stderr,"%s: Warning: '%s' header romsize is 0.\n",argv[0],argv[1]);
		fprintf(stderr,"%s: Warning: This rom will be copied.\n",argv[0]);
		saferomsize=filesize;
	}
	else {
		//Check if rom is big enough to contain the rom
		if (filesize < romsize.data) {
			fprintf(stderr,"%s: Error: '%s' is too small to contain the whole rom (corrupt rom?).\n",argv[0],argv[1]);
			exit(1);
		}
		
		//Check if rom seems to have been trimmed before
		if (filesize == romsize.data) {
			fprintf(stderr,"%s: Warning: '%s' is the same size as the trimmed rom will be.\n",argv[0],argv[1]);
			fprintf(stderr,"%s: Warning: Maybe the rom have already been trimmed?\n",argv[0]);
		}
		
		//Check if rom have a wifi block
		//Wifi data is located on 136 bytes after rom data
		if (filesize >= romsize.data+136) {
			//Read wifi_data from rom
			char wifi_data[136];
			fseek(input,romsize.data,SEEK_SET);
			fread(wifi_data,1,sizeof(wifi_data),input);
			//Compare with 0x00 and 0xFF
			char wifi_compare[136];
			memset(wifi_compare,0x00,136);
			if (memcmp(wifi_data,wifi_compare,136) != 0) {
				//wifi_data is NOT filled with 0x00
				memset(wifi_compare,0xFF,136);
				if (memcmp(wifi_data,wifi_compare,136) != 0) {
					//wifi_data is NOT filled with 0xFF
					//Since wifi_data doesn't consist of 0x00 or 0xFF, the rom contains a wifi block
					wifi_block=1;
					saferomsize+=136;
				}
			}
		}
		else {
			fprintf(stderr,"%s: Warning: '%s' is too small to contain a wifi block.\n",argv[0],argv[1]);
			fprintf(stderr,"%s: Warning: This shouldn't be a problem if the rom have been properly trimmed before.\n",argv[0]);
		}
	}
	
	//Print info
	if (argc < 3 || debug) {
		printf("%s: ROM size: %d bytes.\n",argv[0],romsize.data);
		printf("%s: Safe ROM size: %d bytes.\n",argv[0],saferomsize);
		printf("%s: Wifi block? %s\n",argv[0],wifi_block?"Yes":"No");
		printf("%s: Can save: %d bytes.\n",argv[0],(filesize-saferomsize));
	}
	
	//Output trimmed rom
	if (argc >= 3) {
		//Open output
		#ifdef DEBUG
		printf("%s: Opening output file '%s'.\n",argv[0],argv[2]);
		#endif
		FILE *output;
		if ((output=fopen(argv[2],"wb")) == NULL) {
			fprintf(stderr,"%s: fopen() failed in file %s, line %d.\n",argv[0],__FILE__,__LINE__);
			exit(1);
		}
		
		//Reset input pos
		rewind(input);
		
		//Start copying
		#ifdef DEBUG
		printf("%s: Copying data.\n",argv[0]);
		#endif
		char buffer[BUFFER];
		unsigned int fpos=0;
		unsigned int tocopy=BUFFER;
		while (fpos < saferomsize) {
			if (fpos+BUFFER > saferomsize) {
				tocopy=saferomsize-fpos;
			}
			fread(buffer,tocopy,1,input);
			fwrite(buffer,tocopy,1,output);
			fpos+=tocopy;
		}
		
		//Done
		printf("%s: Trimmed '%s' to %d bytes (saved %.2f MB).\n",argv[0],argv[1],saferomsize,(filesize-saferomsize)/(float)1000000);
		
		//Close output
		#ifdef DEBUG
		printf("%s: Closing output.\n",argv[0]);
		#endif
		if (fclose(output) == EOF) {
			fprintf(stderr,"%s: fclose() failed in file %s, line %d.\n",argv[0],__FILE__,__LINE__);
			exit(1);
		}
	}
	
	//Close input
	#ifdef DEBUG
	printf("%s: Closing input.\n",argv[0]);
	#endif
	if (fclose(input) == EOF) {
		fprintf(stderr,"%s: fclose() failed in file %s, line %d.\n",argv[0],__FILE__,__LINE__);
		exit(1);
	}
	
	return 0;
}

File: ndstrim.rb
#!/usr/bin/env ruby

=begin
This file is part of NDSTrim.

NDSTrim is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

NDSTrim is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with NDSTrim.  If not, see <http://www.gnu.org/licenses/>

----------------------------------------------------------------

Implementation of a 2.5 generation NDS Rom Trimmer
Authors: recover and Azimuth
Excessive documentation

A 1st gen trimmer merely backtraced from the end of the rom looking
for FF or 00 padding and removing it.

A 2nd gen trimmer checks the rom size that's defined in the header
and cuts the rom to that size. This is a much better way of trimming
roms, however, some wifi games have a wifi block which exists after the
rom size in the header and so the wifi block trimmed off with the rest
of the padding.

A 2.5 gen trimmer checks for this wifi block and excludes it
from being trimmed in the case of its existance.
=end

class NDSRom
  attr_reader :file_name, :rom_size, :wifi_block

  def initialize(file)

    @file_name = file

    #First we check the file size, if that is correct we
    #read the rom size in the header. If the rom size in the header is not
    #zero we proceed to reading the wifi block and checking whether or not
    #it is present. The final check before trimming is to check if the rom
    #has already been trimmed.
    check_file_size
    read_rom_size
    read_wifi_block
    check_trimmed
  end

  def check_file_size

    #If the rom size is less that 0x200 the file is the size
    #of the nds rom header, meaning there is no rom data
    if File.size(@file_name) <= 0x200
      raise_error("File too small")
    end
  end

  def read_rom_size

    #seek to 0x80 and read four bytes, this
    #is the location of the rom size in the header
    #rewind back to file start
    @file_name.seek(0x80)
    @rom_size = @file_name.read(4).unpack('I')[0]
    @file_name.rewind

    if @rom_size == 0
      raise_error("Rom Size in Header is zero")
    end
  end

  def read_wifi_block

    #seek to the rom size in the header, read the next 136
    #bytes after this, which is the wifi block
    @file_name.seek(@rom_size)
    @wifi_block = @file_name.read(136)
    @file_name.rewind
    check_wifi_block
  end

  def check_wifi_block

    #The wifi block is 136 bytes long, if it doesn't exist it will be padded like all the
    #data after the rom size in the header. ROMs are usually padded with FF or 00,
    #we check for both these paddings. If the 136 bytes after the rom_size is neither then
    #it must contain a working wifi_block.
    unless @wifi_block == ("\000" * 136) || @wifi_block == ("\377" * 136)
      puts "Wifi block exists"

      #Append wifi block if it exists.
      @rom_size = @rom_size + 136
    else
      return false
    end
  end

  def check_trimmed

    #If the current ROM size is equal to the theoretical ROM
    #size(the rom size in header + wifi block) then the ROM
    #has been (correctly)trimmed before.
    if @rom_size == File.size(@file_name)
      raise_error("Rom has already been trimmed")
    end
  end

  def trim(out_file_name=nil)

    unless @file_name == out_file_name || out_file_name.nil?

      #While the rom's offset is less than the size it should be trimmed to
      #write the number of bytes defined in the buffer from the untrimmed
      #file to the trimmed file. Once the ROM's offset + the buffer is greater
      #than the size it should be the buffer is redefined to be the required rom_size
      #subtracted by the rom offset, which is essentially the bytes left to be written
      #to the new file so that it equals the required size.
      File.open(out_file_name, 'wb') do |rom|
        buffer = 1_000_000
        while rom.pos < @rom_size
          if rom.pos + buffer > @rom_size
            buffer = @rom_size - rom.pos
          end
          rom.write(@file_name.read(buffer))
        end
      end
    else

      #truncates file in place
      @file_name.truncate(@rom_size)
    end
  end

  def raise_error(error)

    #print error to stdout and exit 1
    puts error
    exit 1
  end
end

if __FILE__ == $0

  #Creates a block that passes an open file to
  #the class NDSRom, this way we can avoid
  #re-opening the same file over and over
  File.open(ARGV[0], "r+b") do |file|
    rom = NDSRom.new(file)
    rom.trim(ARGV[1])
  end
end
